File: java/Ex3_1/InsertionSort.java
public class InsertionSort {
    

    public static int[] insertionSort(int[] list)
    {
        for (int i = 1 ; i < list.length;i++)
        {
            int temp = list[i];
            int j = i - 1;

            while(j >= 0 && list[j] > temp)
            {
                list[j+1] = list[j];
                j--;

            }
            list[j+1] = temp;
        }
        return list;
    }
    public static int[] bookInsertionSort(int[] list)
    {
        for(int i = 1 ; i < list.length ;i++)
        {
            int j = 0 ;
            while(list[i] > list[j])
            {
                j++;
            }


            int temp = list[i];
            for(int k = i ; k > j;k--)
            {
                list[k] = list[k-1];
            }
            list[j] = temp;

        }
        return list;
    }
    public static void main(String[] args) {
        
        int[] myList = {3,4,9,5,1,3,9};
        int[] sorted = bookInsertionSort(myList);
        for(int i : sorted)
        {
            System.out.print(i+" ,");
        }
    }
    


    
}

File: cpp/InsertionSort.cpp
#include <iostream>

class InsertionSort
{
public:
    // WRITE YOUR CODE HERE
    void sort(int array[], int size)
    {
        for (int i = 1; i < size; i++)
        {
            int element = array[i];
            int placeToInsert = i - 1;
            while (placeToInsert >= 0 && element < array[placeToInsert])
            {
                array[placeToInsert + 1] = array[placeToInsert];
                placeToInsert--;
            }
            array[placeToInsert + 1] = element;
        }
    }
};

int main()
{
    int array[] = {6, 5, 3, 1, 8, 7, 2, 4};
    int n = sizeof(array) / sizeof(array[0]);

    InsertionSort sorter;
    sorter.sort(array, n);

    std::cout << "Sorted array = {";
    for (int i = 0; i < n; i++)
    {
        std::cout << array[i];
        if (i < n - 1)
        {
            std::cout << ", ";
        }
    }
    std::cout << "}" << std::endl;

    return 0;
}
